### PR TITLE
Fix for broken table sorting feature in table component

### PR DESCRIPTION
**Pull request by MarkvdVorst**
Added types where I could for now in helper service.
In the table settings interface I changed reportMetadata to a Report array instead of any array. 
Because of the change in table settings, a checked field must be added to Report interface.
Moved the sortData function to the table component as it is only being used in that component.
Added a test for the table sorting

### CODE REVIEW

File: src/app/debug/table/table.component.ts [hunk 1]
```diff
@@ -371,9 +373,13 @@ export class TableComponent implements OnInit, OnDestroy {
   }
 
   openReportInTab(): void {
-    let reportTab = this.tableSettings.reportMetadata.find((report) => report.checked);
+    let reportTab: Report | undefined = this.tableSettings.reportMetadata.find((report) => report.checked);
```
Comment (MatthijsSmets) [inline src/app/debug/table/table.component.ts:376]: This let can be a const

File: src/app/debug/table/table.component.ts [hunk 2]
```diff
@@ -662,4 +667,32 @@ export class TableComponent implements OnInit, OnDestroy {
...
+      }
+      return 0;
+    });
+  }
+
+  protected readonly String = String;
+}
+
+function compare(a: number | string, b: number | string, isAsc: boolean): number {
+  return (a < b ? -1 : 1) * (isAsc ? 1 : -1);
```
Comment (MatthijsSmets) [inline src/app/debug/table/table.component.ts:697]: Is it me or does this seem overly complex? Have you looked at [ngBootstrap tables](https://ng-bootstrap.github.io/#/components/table/examples#sortable)?
Also, why is this function outside of the component class?

File: src/app/shared/services/helper.service.ts [hunk 3]
```diff
@@ -161,14 +129,14 @@ export class HelperService {
...
-  createChildNodes(rootNode: any, index: number, parentMap: any[]) {
+  createChildNodes(rootNode: any, index: number, parentMap: any[]): void {
     let previousNode = rootNode;
     let checkpoints: any[] = previousNode.value.checkpoints;
 
     if (checkpoints && checkpoints.length > 0) {
       for (let checkpoint of checkpoints) {
         const img: string = this.getImage(checkpoint.type, checkpoint.encoding, checkpoint.level);
-        let showingId = this.getCheckpointOrStorageId(checkpoint, false);
+        let showingId: string = this.getCheckpointOrStorageId(checkpoint, false);
```
Comment (MatthijsSmets) [inline src/app/shared/services/helper.service.ts:139]: let can be a const

File: src/app/shared/services/helper.service.ts [hunk 4]
```diff
@@ -128,14 +96,14 @@ export class HelperService {
...
     this.createChildNodes(rootNode, index, parentMap);
     return rootNode;
   }
 
   createNode(report: Report, showingId: string, icon: string, index: number, level: number) {
-    let expanded = true;
+    let expanded: boolean = true;
     if (level > 0) {
       expanded = false;
     }
```
Comment (MatthijsSmets) [inline src/app/shared/services/helper.service.ts:109]:
```suggestion
    const expanded: boolean = level <= 0;
```

File: src/app/shared/services/helper.service.ts [hunk 5]
```diff
@@ -128,14 +96,14 @@ export class HelperService {
     let index: number = 0;
     let parentMap: any[] = [];
 
-    let showingId = this.getCheckpointOrStorageId(report, true);
+    let showingId: string = this.getCheckpointOrStorageId(report, true);
```
Comment (MatthijsSmets) [inline src/app/shared/services/helper.service.ts:99]: let can be a const
parentMap can also be a const
Also this function should be refactored, so I dont understand why you've added types here

File: src/app/debug/table/table.component.ts [hunk 6]
```diff
@@ -662,4 +667,32 @@ export class TableComponent implements OnInit, OnDestroy {
...
+      for (const [index, element] of headersA.entries()) {
+        if (Number(sort.active) === index) {
+          return compare(element[1], headersB[index][1], isAsc);
+        }
+      }
+      return 0;
+    });
+  }
+
+  protected readonly String = String;
```
Comment (MatthijsSmets) [inline src/app/debug/table/table.component.ts:693]: ?

File: src/app/debug/table/table.component.ts [hunk 7]
```diff
@@ -371,9 +373,13 @@ export class TableComponent implements OnInit, OnDestroy {
...
   openReportInTab(): void {
-    let reportTab = this.tableSettings.reportMetadata.find((report) => report.checked);
+    let reportTab: Report | undefined = this.tableSettings.reportMetadata.find((report) => report.checked);
+    if (!reportTab) {
+      this.toastService.showDanger('Could not find report', 'No report found that was selected.');
+      return;
+    }
     this.httpService
-      .getReport(reportTab.storageId, this.viewSettings.currentView.storageName)
+      .getReport(String(reportTab.storageId), this.viewSettings.currentView.storageName)
```
Comment (MatthijsSmets) [inline src/app/debug/table/table.component.ts:382]: I see you casting storage id to a string a lot of times, I also see you've change the type in the interface from a number to a string, either change the method input parameters to accept a string, or keep it a number.
I think that if it can be a number then it should be a number.

**Review by MatthijsSmets — CHANGES_REQUESTED** [review]

File: src/app/debug/table/table.component.ts (hunk 1 above)
Comment (MarkvdVorst) [reply src/app/debug/table/table.component.ts]: added the change

File: src/app/debug/table/table.component.ts (hunk 2 above)
Comment (MarkvdVorst) [reply src/app/debug/table/table.component.ts]: > Also, why is this function outside of the component class?

It was a copy and paste from helper service and angular docs example showed the same thing. I changed it so it is now in the component.

> Is it me or does this seem overly complex? Have you looked at [ngBootstrap tables](https://ng-bootstrap.github.io/#/components/table/examples#sortable)?

I took a look at it and I find that more unreadable that the current code, but I might be biased since I just worked on it. The example for sorting seems like you would write a lot of code that the matSort already does for you. Would be great to hear your opinion on this.

File: src/app/shared/services/helper.service.ts (hunk 3 above)
Comment (MarkvdVorst) [reply src/app/shared/services/helper.service.ts:136]: added the change

File: src/app/shared/services/helper.service.ts (hunk 4 above)
Comment (MarkvdVorst) [reply src/app/shared/services/helper.service.ts]: Added the change

File: src/app/shared/services/helper.service.ts (hunk 5 above)
Comment (MarkvdVorst) [reply src/app/shared/services/helper.service.ts]: I added the change.

> Also this function should be refactored, so I dont understand why you've added types here

I was just going through the helper service adding types wherever it was easy to add it so we don't have to later on.

File: src/app/debug/table/table.component.ts (hunk 6 above)
Comment (MarkvdVorst) [reply src/app/debug/table/table.component.ts]: Removed this line.
My IDE added this when I tried converting index in the template to string. I now changed it so the sort works based on the header.
This did mean I had to add another method which converted the metadata label header to the metadataname field of the report type.

File: src/app/debug/table/table.component.ts (hunk 7 above)
Comment (MarkvdVorst) [reply src/app/debug/table/table.component.ts]: I tried getting rid of the string casting at all places where I could, but removing the string casting in tab.service.ts seems to break the debug_fileupload E2E test and I'm still not sure why at this point so I opted to leave it in that service for now.

File: src/app/debug/table/table.component.ts (hunk 2 above)
Comment (MatthijsSmets) [reply src/app/debug/table/table.component.ts]: It might be easier to just use [angular material table](https://material.angular.io/components/table/overview#sorting)

Comment (MarkvdVorst) [reply src/app/debug/table/table.component.ts]: @MatthijsSmets Replaced the old table with the angular material table. This resulted in not needing a sorting method anymore.

**Comment by MarkvdVorst** [comment]
Tbody element is now generated by the Angular Material Table and because of that I had to remove the data-cy attribute from the element. Because of this an extra get method was needed for the cypress test that required to look into the tbody.

File: cypress/e2e/debug_download.cy.js [hunk 8]
```diff
@@ -89,7 +89,7 @@ describe('Debug tab download', function () {
   // TODO: Fix issue and re-enable test.
   xit('Download all open reports', function () {
     const downloadsFolder = Cypress.config('downloadsFolder');
-    cy.get('[data-cy-debug="tableBody"]').find('tr').should('have.length', 2);
+    cy.get('[data-cy-debug="tableBody"]').get('tbody').find('tr').should('have.length', 2);
```
Comment (MatthijsSmets) [inline cypress/e2e/debug_download.cy.js:92]: Theres a lot of places where you check the length of the table, to follow the [DRY rule](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), make it a function:
```ts
Cypress.Commands.add('checkDebugTableLength', (n) => {
  if (n === 0){
    cy.get('[data-cy-debug="tableBody"] tbody').find('tr').should('not.exist');
  } else {
    cy.get('[data-cy-debug="tableBody"] tbody').find('tr').should('have.length', n);
  }
});

```

File: cypress/e2e/testtab.cy.js [hunk 9]
```diff
@@ -8,13 +8,13 @@ describe('About the Test tab', () => {
     cy.createOtherReport();
     cy.initializeApp();
     const storageIds = [];
-    cy.get('[data-cy-debug="tableBody"] tr').each($row => {
+    cy.get('[data-cy-debug="tableBody"] tbody tr').each($row => {
```
Comment (MatthijsSmets) [inline cypress/e2e/testtab.cy.js:11]: Previously you've done: 
```ts
cy.get('[data-cy-debug="tableBody"]').get('tbody').find('tr');
```
why do it differently here, please be consistent.

File: src/app/app.component.ts [hunk 10]
```diff
@@ -110,11 +110,11 @@ export class AppComponent implements OnInit, OnDestroy, AfterViewInit {
   }
 
   openReportInSeparateTab(data: ReportData): void {
-    const tabIndex: number = this.tabs.findIndex((tab: Tab): boolean => tab.id === data.report.storageId);
+    const tabIndex: number = this.tabs.findIndex((tab: Tab): boolean => tab.id === String(data.report.storageId));
```
Comment (MatthijsSmets) [inline src/app/app.component.ts:113]: Maybe it's best if `id` in `Tab` is also of type number

File: src/app/debug/table/table.component.html [hunk 11]
```diff
@@ -79,7 +79,7 @@ <h4 class="progress-report-timout">[One or more reports are in progress for more
 
       @switch (getAmountOfReportsSelected()) {
         @case (1) {
-          <button data-cy-debug="openReportTab" class="btn btn-info btn-table" title="Open In New Tab" (click)="openReportInTab()">
+          <button class="btn btn-info btn-table" title="Open In New Tab" (click)="openReportInTab()">
```
Comment (MatthijsSmets) [inline src/app/debug/table/table.component.html:82]: Why did you remove the `data-cy` tag here?

File: src/app/debug/table/table.component.html [hunk 12]
```diff
@@ -98,42 +98,43 @@ <h4 class="progress-report-timout">[One or more reports are in progress for more
...
+
+          <!--Values are passed to their respective header and row elements-->
+          <tr mat-header-row *matHeaderRowDef="getDisplayedColumnNames(viewSettings.currentView.metadataLabels)"></tr>
+          <tr mat-row *matRowDef="let row; let rowIndex = index; columns: getDisplayedColumnNames(viewSettings.currentView.metadataLabels)"
+              class="table-row"
+              [ngClass]="{'highlight': row.storageId === selectedReportStorageId}"
+              [attr.data-cy-record-table-index]="rowIndex"
+              [style.background-color]="getStatusColor(row)"
+              (click)="openSelectedReport(row.storageId);"
+          ></tr>
```
Comment (MatthijsSmets) [inline src/app/debug/table/table.component.html:137]: I don't understand why you've split the table into three parts like that?

File: src/app/debug/table/table.component.ts [hunk 13]
```diff
@@ -419,7 +425,7 @@ export class TableComponent implements OnInit, OnDestroy {
 
   compareTwoReports(): void {
     let compareReports: any = {};
-    let selectedReports: string[] = this.tableSettings.reportMetadata
+    let selectedReports: number[] = this.tableSettings.reportMetadata
```
Comment (MatthijsSmets) [inline src/app/debug/table/table.component.ts:428]: This can be a const

File: src/app/shared/services/helper.service.ts [hunk 14]
```diff
@@ -176,8 +180,8 @@ export class HelperService {
     parent.items.push(node);
   }
 
-  getSelectedIds(reports: any[]): string[] {
-    let copiedIds: string[] = [];
+  getSelectedIds(reports: any[]): number[] {
+    let copiedIds: number[] = [];
     this.getSelectedReports(reports).forEach((report: Report) => copiedIds.push(report.storageId));
```
Comment (MatthijsSmets) [inline src/app/shared/services/helper.service.ts:185]: While you're at it, can you refactor this to use a a `for...of` loop instead of a `foreach` loop?

**Review by MatthijsSmets — CHANGES_REQUESTED** [review]
Looks good, just some minor things

File: src/app/app.component.ts (hunk 10 above)
Comment (MarkvdVorst) [reply src/app/app.component.ts:113]: I'm afraid that's gonna have to get it's own PR for a different refactor since this PR will probably get too big otherwise. The createCompareTabId method from the helperService returns an odd tabId that can't be converted to a number
```
createCompareTabId(originalReport: Report, runResultReport: Report): string {
    return `${originalReport.storageId}-${runResultReport.storageId}`;
 }
  ```
  I can't take a quick solution for this for now so maybe it's best to add it as an issue to the refactor board. What do you think?

File: src/app/debug/table/table.component.html (hunk 12 above)
Comment (MarkvdVorst) [reply src/app/debug/table/table.component.html:137]: I followed the example code from the angular docs on the Material Table. The only big thing that I changed from the example code is that this creates headers dynamically depending on the labels in the viewSettings.currentView.metadataLabels

File: cypress/e2e/testtab.cy.js (hunk 9 above)
Comment (MarkvdVorst) [reply cypress/e2e/testtab.cy.js]: The cypress tests in general are pretty inconsistant for this. I changed it to a new command call, but it might be worth refactoring the cypress tests for this at some point.

File: src/app/debug/table/table.component.html (hunk 11 above)
Comment (MarkvdVorst) [reply src/app/debug/table/table.component.html]: Woops. I don't know how that happened actually. I added it back.

File: src/app/shared/services/helper.service.ts (hunk 14 above)
Comment (MarkvdVorst) [reply src/app/shared/services/helper.service.ts]: Added the change.

File: src/app/app.component.ts (hunk 10 above)
Comment (MatthijsSmets) [reply src/app/app.component.ts:113]: We can leave it for now.

File: cypress/e2e/debug_download.cy.js (hunk 8 above)
Comment (MarkvdVorst) [reply cypress/e2e/debug_download.cy.js]: added a global function that asserts the length of the tbody for n length now.

File: cypress/support/functions.ts [hunk 15]
```diff
@@ -1,19 +1,25 @@
...
+          cy.wrap($reportRow).find('[type=checkbox]').uncheck();
+        }
+        cy.get('[data-cy-test="reportChecked"]:checked').should('have.length', 1);
+        cy.wait(1000);
+      });
+  },
+
+  assertDebugTableLength: (length: number) => {
+    cy.getTableBody().find('tr').should('have.length', length);
+  },
```
Comment (MatthijsSmets) [inline cypress/support/functions.ts:24]: Please move this to commands.js

File: cypress/e2e/debug_download.cy.js [hunk 16]
```diff
@@ -52,7 +52,7 @@ describe('Debug tab download', function () {
           });
         cy.clearDebugStore();
         cy.get('[data-cy-debug="refresh"]').click();
-        cy.get('[data-cy-debug="tableBody"]').find('tr').should('not.exist');
+        cy.getTableBody().find('tr').should('not.exist');
```
Comment (MatthijsSmets) [inline cypress/e2e/debug_download.cy.js:55]: Use your function for this:
```suggestion
        cy.functions.assertDebugTableLength(0);
```

File: cypress/e2e/debug_filterTable.cy.js [hunk 17]
```diff
@@ -13,21 +13,21 @@ describe('Table size and toggle filter', () => {
 
   it('Typing in a table size and retyping it', () => {
     // We only assume here that the default is two or more.
-    cy.get('[data-cy-debug="tableBody"]').find('tr').should('have.length', 2);
+    cy.functions.assertDebugTableLength(2);
     cy.get('[data-cy-debug="displayAmount"]').type('{selectAll}{del}{enter}');
-    cy.get('[data-cy-debug="tableBody"]').find('tr').should('not.exist');
+    cy.getTableBody().find('tr').should('not.exist');
```
Comment (MatthijsSmets) [inline cypress/e2e/debug_filterTable.cy.js:18]: Here as well:
```suggestion
    cy.functions.assertDebugTableLength(2);
```

File: cypress/e2e/debug_filterTable.cy.js [hunk 18]
```diff
@@ -13,21 +13,21 @@ describe('Table size and toggle filter', () => {
...
     // From now on, we type one character at a time. Cypress can type very rapidly.
     // We do not expect our app to catch up without guards.
     cy.get('[data-cy-debug="displayAmount"]').type('{selectAll}1{enter}');
-    cy.get('[data-cy-debug="tableBody"]').find('tr').should('have.length', 1);
+    cy.functions.assertDebugTableLength(1);
     cy.get('[data-cy-debug="displayAmount"]').type('{backspace}{enter}');
-    cy.get('[data-cy-debug="tableBody"]').find('tr').should('not.exist');
+    cy.getTableBody().find('tr').should('not.exist');
```
Comment (MatthijsSmets) [inline cypress/e2e/debug_filterTable.cy.js:24]: Here as well:
```suggestion
    cy.functions.assertDebugTableLength(0);
```

File: cypress/e2e/debug_filterTable.cy.js [hunk 19]
```diff
@@ -13,21 +13,21 @@ describe('Table size and toggle filter', () => {
...
     cy.get('[data-cy-debug="displayAmount"]').type('{selectAll}2{enter}');
-    cy.get('[data-cy-debug="tableBody"]').find('tr').should('have.length', 2);
+    cy.functions.assertDebugTableLength(2);
     cy.get('[data-cy-debug="displayAmount"]').type('{backspace}{enter}');
-    cy.get('[data-cy-debug="tableBody"]').find('tr').should('not.exist');
+    cy.getTableBody().find('tr').should('not.exist');
```
Comment (MatthijsSmets) [inline cypress/e2e/debug_filterTable.cy.js:28]: Here as well:
```suggestion
    cy.functions.assertDebugTableLength(0);
```

File: cypress/e2e/debug_refresh.cy.js [hunk 20]
```diff
@@ -8,11 +8,11 @@ describe('Refresh', () => {
   afterEach(() => cy.resetApp());
 
   it('New reports are only shown on refresh', () => {
-    cy.get('[data-cy-debug="tableBody"]').find('tr').should('not.exist');
+    cy.getTableBody().find('tr').should('not.exist');
```
Comment (MatthijsSmets) [inline cypress/e2e/debug_refresh.cy.js:11]: Here as well:
```suggestion
    cy.functions.assertDebugTableLength(0);
```

File: cypress/e2e/debug_reportGenerator.cy.js [hunk 21]
```diff
@@ -12,11 +12,11 @@ describe('Report generator', () => {
   afterEach(() => cy.resetApp());
 
   it('disable and enable', () => {
-    cy.get('[data-cy-debug="tableBody"]').find('tr').should('not.exist');
+    cy.getTableBody().find('tr').should('not.exist');
```
Comment (MatthijsSmets) [inline cypress/e2e/debug_reportGenerator.cy.js:15]: Here as well:
```suggestion
    cy.functions.assertDebugTableLength(0);
```

File: cypress/e2e/debug_toast.cy.js [hunk 22]
```diff
@@ -6,11 +6,11 @@ describe('Test toast window', () => {
   afterEach(() => cy.resetApp());
 
   it('When new report appears in table then toast window shown', () => {
-    cy.get('[data-cy-debug="tableBody"]').find('tr').should('not.exist');
+    cy.getTableBody().find('tr').should('not.exist');
```
Comment (MatthijsSmets) [inline cypress/e2e/debug_toast.cy.js:9]: Here as well:
```suggestion
    cy.functions.assertDebugTableLength(0);
```

File: cypress/e2e/transformation.cy.js [hunk 23]
```diff
@@ -28,7 +28,7 @@ describe('Report transformation', () => {
     cy.createOtherReport();
     cy.get('[data-cy-debug="refresh"]').click();
     cy.wait(100);
-    cy.get('[data-cy-debug="tableBody"]')
+    cy.getTableBody()
       .find('tr')
       .should('have.length', 1)
       .click();
```
Comment (MatthijsSmets) [inline cypress/e2e/transformation.cy.js:34]: Here as well:
```suggestion
    cy.functions.assertDebugTableLength(1);
```

File: cypress/support/functions.ts (hunk 15 above)
Comment (MatthijsSmets) [inline cypress/support/functions.ts:24]:
```suggestion
  assertDebugTableLength: (length: number) => {
    if (length === 0) {
      cy.getTableBody().find('tr').should('not.exist');
    } else {
      cy.getTableBody().find('tr').should('have.length', length);
    }
  },
```

File: src/app/debug/debug-tree/debug-tree.component.ts [hunk 24]
```diff
@@ -90,7 +90,7 @@ export class DebugTreeComponent implements OnDestroy {
     for (let report of reports) {
       if (report.storageName === currentView.storageName) {
         this.httpService
-          .getUnmatchedCheckpoints(report.storageName, report.storageId.toString(), currentView.currentViewName)
+          .getUnmatchedCheckpoints(report.storageName, report.storageId, currentView.currentViewName)
```
Comment (MatthijsSmets) [inline src/app/debug/debug-tree/debug-tree.component.ts:93]:
```suggestion
          .getUnmatchedCheckpoints(report.storageName, report.storageId, currentView.name)
```

File: src/app/debug/table/table.component.ts [hunk 25]
```diff
@@ -183,6 +191,7 @@ export class TableComponent implements OnInit, OnDestroy {
         next: (value: Report[]) => {
           this.setUniqueOptions(value);
           this.tableSettings.reportMetadata = value;
+          this.tableDataSource.data = [...value];
```
Comment (MatthijsSmets) [inline src/app/debug/table/table.component.ts:194]: Would this not work? Using the spread operator loops over the list which is unecessary if the resulting list doesn't contain anything other than the list
```suggestion
          this.tableDataSource.data = value;
```

**Review by MatthijsSmets — CHANGES_REQUESTED** [review]

File: cypress/support/functions.ts (hunk 15 above)
Comment (MarkvdVorst) [reply cypress/support/functions.ts]: added the change

File: cypress/e2e/debug_download.cy.js (hunk 16 above)
Comment (MarkvdVorst) [reply cypress/e2e/debug_download.cy.js]: Added the change.

File: cypress/e2e/debug_filterTable.cy.js (hunk 17 above)
Comment (MarkvdVorst) [reply cypress/e2e/debug_filterTable.cy.js]: Added the change.

File: cypress/e2e/debug_filterTable.cy.js (hunk 18 above)
Comment (MarkvdVorst) [reply cypress/e2e/debug_filterTable.cy.js]: Added the change.

File: cypress/e2e/debug_filterTable.cy.js (hunk 19 above)
Comment (MarkvdVorst) [reply cypress/e2e/debug_filterTable.cy.js]: Added the change.

File: cypress/e2e/debug_refresh.cy.js (hunk 20 above)
Comment (MarkvdVorst) [reply cypress/e2e/debug_refresh.cy.js]: Added the change.

File: cypress/e2e/debug_reportGenerator.cy.js (hunk 21 above)
Comment (MarkvdVorst) [reply cypress/e2e/debug_reportGenerator.cy.js]: Added the change.

File: cypress/e2e/debug_toast.cy.js (hunk 22 above)
Comment (MarkvdVorst) [reply cypress/e2e/debug_toast.cy.js]: Added the change.

File: cypress/e2e/transformation.cy.js (hunk 23 above)
Comment (MarkvdVorst) [reply cypress/e2e/transformation.cy.js]: Added the change.

File: cypress/support/functions.ts (hunk 15 above)
Comment (MarkvdVorst) [reply cypress/support/functions.ts]: Added the change. Conditional branching was suggested by ESLint so I used that instead.

File: src/app/debug/debug-tree/debug-tree.component.ts (hunk 24 above)
Comment (MarkvdVorst) [reply src/app/debug/debug-tree/debug-tree.component.ts]: Added the change.

File: src/app/debug/table/table.component.ts (hunk 25 above)
Comment (MarkvdVorst) [reply src/app/debug/table/table.component.ts]: Added the change. It would work with or without. I was more afraid of shared mutable state and that the material table would change the order of the original values. Turns out it does not.

**Review by MatthijsSmets — APPROVED** [review]

**Review by MatthijsSmets — APPROVED** [review]
